llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -265,11 +265,13 @@ bool CombinerHelper::matchFreezeOfSingleMaybePoisonOperand(
     }
   }
 
+  Observer.changingInstr(*OrigDef);
   cast<GenericMachineInstr>(OrigDef)->dropPoisonGeneratingFlags();
```

Comment:
This seems like a problem, the match function should not be modifying the IR

Reply:
In Instcombine, the change is *above* the returns.
https://github.com/llvm/llvm-project/blob/cb7995a1308c1bef396f9a2b2b14472cbdcca755/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L4428
We could move the change into both `MatchInfo` s.